src/Features/LanguageServer/Protocol/Handler/Diagnostics/DiagnosticSources/ProjectDiagnosticSource.cs

```diff
@@ -11,28 +11,59 @@
 
 namespace Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
 
-internal sealed record class ProjectDiagnosticSource(Project Project, Func<DiagnosticAnalyzer, bool>? ShouldIncludeAnalyzer) : IDiagnosticSource
+internal abstract class ProjectDiagnosticSource : IDiagnosticSource
```

Comment:
```suggestion
internal abstract class AbstractProjectDiagnosticSource : IDiagnosticSource
```

also, i'd keep with primary constructors.